File: intrinsic/skills/internal/get_footprint_context_impl.cc
// Copyright 2023 Intrinsic Innovation LLC

#include "intrinsic/skills/internal/get_footprint_context_impl.h"

#include <memory>
#include <string>
#include <utility>

#include "absl/log/log.h"
#include "absl/status/statusor.h"
#include "absl/strings/string_view.h"
#include "intrinsic/resources/proto/resource_handle.pb.h"
#include "intrinsic/skills/proto/equipment.pb.h"
#include "intrinsic/util/status/status_macros.h"
#include "intrinsic/world/objects/frame.h"
#include "intrinsic/world/objects/kinematic_object.h"
#include "intrinsic/world/objects/object_world_ids.h"
#include "intrinsic/world/objects/world_object.h"

namespace intrinsic {
namespace skills {

absl::StatusOr<world::KinematicObject>
GetFootprintContextImpl::GetKinematicObjectForEquipment(
    absl::string_view equipment_name) {
  INTR_ASSIGN_OR_RETURN(const intrinsic_proto::resources::ResourceHandle handle,
                        equipment_.GetHandle(equipment_name));
  return object_world().GetKinematicObject(handle);
}

absl::StatusOr<world::WorldObject>
GetFootprintContextImpl::GetObjectForEquipment(
    absl::string_view equipment_name) {
  INTR_ASSIGN_OR_RETURN(const intrinsic_proto::resources::ResourceHandle handle,
                        equipment_.GetHandle(equipment_name));
  return object_world().GetObject(handle);
}

absl::StatusOr<world::Frame> GetFootprintContextImpl::GetFrameForEquipment(
    absl::string_view equipment_name, absl::string_view frame_name) {
  INTR_ASSIGN_OR_RETURN(const intrinsic_proto::resources::ResourceHandle handle,
                        equipment_.GetHandle(equipment_name));
  return object_world().GetFrame(handle, FrameName(frame_name));
}

}  // namespace skills
}  // namespace intrinsic

File: intrinsic/skills/internal/get_footprint_context_impl.py
# Copyright 2023 Intrinsic Innovation LLC

"""GetFootprintContext implementation provided by the skill service."""

from intrinsic.motion_planning import motion_planner_client
from intrinsic.skills.proto import equipment_pb2
from intrinsic.skills.python import get_footprint_context
from intrinsic.world.python import object_world_client
from intrinsic.world.python import object_world_ids
from intrinsic.world.python import object_world_resources


class GetFootprintContextImpl(get_footprint_context.GetFootprintContext):
  """GetFootprintContext implementation provided by the skill service.

  It is provided by the skill service to a skill and allows access to the world
  and other services that a skill may use.

  Attributes:
    motion_planner: A client for the motion planning service.
    object_world: A client for interacting with the object world.
  """

  @property
  def motion_planner(self) -> motion_planner_client.MotionPlannerClient:
    return self._motion_planner

  @property
  def object_world(self) -> object_world_client.ObjectWorldClient:
    return self._object_world

  def __init__(
      self,
      motion_planner: motion_planner_client.MotionPlannerClient,
      object_world: object_world_client.ObjectWorldClient,
      resource_handles: dict[str, equipment_pb2.ResourceHandle],
  ):
    self._motion_planner = motion_planner
    self._object_world = object_world
    self._resource_handles = resource_handles

  def get_frame_for_equipment(
      self, equipment_name: str, frame_name: object_world_ids.FrameName
  ) -> object_world_resources.Frame:
    return self.object_world.get_frame(
        frame_name, self._resource_handles[equipment_name]
    )

  def get_kinematic_object_for_equipment(
      self, equipment_name: str
  ) -> object_world_resources.KinematicObject:
    return self.object_world.get_kinematic_object(
        self._resource_handles[equipment_name]
    )

  def get_object_for_equipment(
      self, equipment_name: str
  ) -> object_world_resources.WorldObject:
    return self.object_world.get_object(self._resource_handles[equipment_name])
